include/zephyr/drivers/adc/adc_current_sense_shunt.h

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright (c) 2023 The ChromiumOS Authors
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_ADC_CURRENT_SENSE_SHUNT_H_
+#define ZEPHYR_INCLUDE_DRIVERS_ADC_CURRENT_SENSE_SHUNT_H_
+
+#include <zephyr/drivers/adc.h>
+
+struct adc_dt_current_sense_shunt_cfg {
+	const struct adc_dt_spec port;
+	int shunt_micro_ohms;
+	int shunt_gain;
+};
+
+#define ADC_DT_CURRENT_SENSE_SHUNT_CFG_GET_BY_IDX(node_id, idx)                                    \
+	{                                                                                          \
+		.port = ADC_DT_SPEC_GET(node_id),                                                  \
+		.shunt_micro_ohms = DT_PROP(node_id, shunt_resistor_micro_ohms),                   \
+		.shunt_gain = DT_PROP(node_id, gain),                                              \
+	}
+
+#define ADC_DT_CURRENT_SENSE_SHUNT_CFG_GET(node_id)                                                \
+	ADC_DT_CURRENT_SENSE_SHUNT_CFG_GET_BY_IDX(node_id, 0)
+
+static inline int adc_scale_current_dt(const struct adc_dt_current_sense_shunt_cfg *cfg,
+				       int32_t *v_to_c, int32_t offset_mv)
+{
+	/* multiplies by 1,000,000 before dividing by shunt resistance
+	 * in micro-ohms.
+	 */
+	*v_to_c = (*v_to_c - offset_mv) * 1000000 / cfg->shunt_micro_ohms / cfg->shunt_gain;
+
+	return 0;
+}
+
+static inline int adc_get_current_dt(const struct adc_dt_current_sense_shunt_cfg *cfg, int32_t *mA)
+{
+	int ret;
+	int32_t sample_buffer = 0;
+
+	/* Structure defining an ADC sampling sequence */
+	struct adc_sequence sequence = {
+		.buffer = &sample_buffer,
+		/* buffer size in bytes, not number of samples */
+		.buffer_size = sizeof(sample_buffer),
+	};
+	adc_sequence_init_dt(&cfg->port, &sequence);
+
+	ret = adc_read(cfg->port.dev, &sequence);
+	if (ret != 0) {
+		return ret;
+	}
+
+	*mA = sample_buffer;
+	ret = adc_raw_to_millivolts_dt(&cfg->port, mA);
+	if (ret != 0) {
+		return ret;
+	}
+
+	ret = adc_scale_current_dt(cfg, mA, cfg->port.vref_mv / 2);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return 0;
+}
```

Comment:
I would live this out of this header file, there's few ways of setting up the adc channels depending on what the hw supports, just keep the struct, initializer and scaling function here.